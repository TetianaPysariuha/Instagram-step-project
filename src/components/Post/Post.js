import React, { memo } from 'react';
import PropTypes from 'prop-types';
import styles from './Post.module.scss';
import CommentForm from '../CommentForm/CommentForm';

function Post(props) {
  const {
    postId, userName, mainImg, title, description, isLiked, handleCklickLike, handleCklickComments, isFavorite, handleCklickFavorite, comments, isMore, handleCklickShowMore,
  } = props;

  return (
    <div key={postId} className={styles.postContainier}>
      <div className={styles.postWrapper}>
        {userName}
        <img className={styles.maneFoto} src={mainImg} onDoubleClick={handleCklickLike} alt={title} />
        <div className={styles.buttonBox}>
          <div className={styles.groupBtn}>
            <img src={isLiked ? './images/heart-red.svg' : './images/heart-white.svg'} onClick={handleCklickLike} alt={isLiked ? 'liked' : 'not liked'} data-testid="likeBtn" />
            <img src="./images/comment.svg" onClick={handleCklickComments} alt="see all comments" data-testid="commentBtn" />
          </div>
          <img src={isFavorite ? './images/favorite-marked.svg' : './images/favorite.svg'} onClick={handleCklickFavorite} alt="favorite" data-testid="favoriteBtn" />
        </div>
        <p className={styles.description}>{description}</p>
        <div className={styles.commentBox}>
          <div className={styles.comments}>
            {comments}
            <button type="button" className={styles.showMoreBtn} onClick={handleCklickShowMore}>{isMore ? '-----Show more comments-----' : '-----Show less comments-----'}</button>
          </div>
          <div className={styles.addComment}>
            <CommentForm postId={postId} />
          </div>
        </div>
      </div>
    </div>
  );
}

Post.prototype = {
  postId: PropTypes.string.isRequired,
  userName: PropTypes.element.isRequired,
  mainImg: PropTypes.string.isRequired,
  title: PropTypes.string,
  description: PropTypes.string,
  isLiked: PropTypes.bool,
  handleCklickLike: PropTypes.func,
  handleCklickComments: PropTypes.func,
  isFavorite: PropTypes.bool,
  handleCklickFavorite: PropTypes.func,
  comments: PropTypes.element,
  isMore: PropTypes.bool,
  handleCklickShowMore: PropTypes.func,
};

Post.defaultProps = {
  title: '',
  description: '',
  isLiked: false,
  handleCklickLike: () => {},
  handleCklickComments: () => {},
  isFavorite: false,
  handleCklickFavorite: () => {},
  comments: <> </>,
  isMore: false,
  handleCklickShowMore: () => {},
};

export default memo(Post);
