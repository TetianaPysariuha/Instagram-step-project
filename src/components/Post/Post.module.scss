.postContainier{
    width: 60%;
    border: 3px solid rgba($color: gray, $alpha: 0.3);
    margin: 30px auto;
}
.postWrapper{
    width: 100%;
    margin: auto;
}
.maneFoto{
    width: 100%;
    margin: 10px auto;
}
.buttonBox{
    width: 100%;
    display: flex;
    justify-content: space-between;
    padding: 10px;
    box-sizing:border-box;
    & img{
        cursor: pointer;
    }
}
.groupBtn{
    display: flex;
    justify-content: space-around;
    width: 60px;
}
.description{
    width: 100%;
}
.commentBox{
    width: 100%;
    display: flex;
    flex-direction: column;
}
.comments{
    width: 100%;
    display: flex;
    flex-direction: column;
    margin: 10px;
    font-weight: 400;
    &>ul{
        text-decoration: none;
        list-style: none;
    }
    & span{
        font-weight: 700;
    }
}
.showMoreBtn{
    border: none;
    background-color: white;
    cursor: pointer;
    padding: 20px;
    align-self: center;
}
.addComment{
border-top: 1px solid rgba($color: gray, $alpha: 0.1);
}
