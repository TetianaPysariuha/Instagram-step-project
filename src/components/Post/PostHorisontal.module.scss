.postContainier {
    width: 100%;
    height: 85vh;
    /*     border: 3px solid rgba($color: gray, $alpha: 0.3); */
    margin: auto;
    border-radius: 10px;
    max-height: 90vh;
}

.postWrapper {
    height: 100%;
    display: grid;
    grid-template-columns: 1fr 400px;
    grid-template-rows: 30px 1fr 3fr 50px 50px;
}

.maneFoto {
    width: 100%;
    height: 100%;
    margin: auto;
    grid-column-start: 1;
    grid-row-start: 1;
    grid-row-end: 6;
    background-color: black;
    display: flex;
    align-items: center;

    &>img {
        display: block;
    }

    background-image: none !important;
}

.maneFoto img {
    max-height: 85vh;
    width: 100%;
    object-fit: contain;
}

.buttonBox {
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: space-between;
    padding: 10px;
    box-sizing: border-box;

    & img {
        cursor: pointer;
    }
}

.buttonBox img {
    width: 25px;
}

.groupBtn {
    display: flex;
    justify-content: space-around;
    width: 70px;
}

.description {
    margin: 20px;
    padding: 10px;
    grid-row-start: 2;
    grid-row-end: 3;
    border-top: 1px solid rgba($color: gray, $alpha: 0.3);
}

.comments {
    display: flex;
    flex-direction: column;
    max-width: 90%;
    margin: 10px;
    font-weight: 400;

    &>ul {
        text-decoration: none;
        list-style: none;
    }

    & span {
        font-weight: 700;
    }

    grid-row-start: 3;
    grid-row-end: 4;
    margin: 20px;
    overflow-y: scroll;
    scrollbar-width: none;
}

.comments li {
    width: 100%;
}

.showMoreBtn {
    border: none;
    background-color: white;
    cursor: pointer;
    padding: 20px;
    align-self: center;

}

.addComment {
    border-top: 1px solid rgba($color: gray, $alpha: 0.1);
    width: 100%;
    grid-row-start: 5;
    grid-row-end: 6;
    justify-self: center;
}