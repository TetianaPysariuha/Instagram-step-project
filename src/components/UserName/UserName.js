import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import styles from './UserName.module.scss';
import defaultAvatar from './default-avatar.jpg';

function UserName({
  image, nickname, additionalString, comment, additionalStringFunc,
}) {
  return (
    <div className={styles.userNameContainer}>
      <Link to={`/${nickname}/`} className={styles.avatarLink}>
        <img src={image} alt="avatar" className={styles.avatar} />
      </Link>
      <div className={styles.nicknameWrraper}>
        <div>
          <Link to={`/${nickname}/`} className={styles.nickname}>
            <span>{nickname}</span>
          </Link>
          <p className={styles.comment}>{comment}</p>
        </div>
        {additionalString && <p className={styles.additionalString} onClick={additionalStringFunc}>{additionalString}</p>}
      </div>
    </div>
  );
}

UserName.prototype = {
  image: PropTypes.string,
  nickname: PropTypes.string.isRequired,
  additionalString: PropTypes.string,
  comment: PropTypes.string,
  additionalStringFunc: PropTypes.func,
};

UserName.defaultProps = {
  image: defaultAvatar,
  additionalString: '',
  additionalStringFunc: () => {},
};

export default UserName;
