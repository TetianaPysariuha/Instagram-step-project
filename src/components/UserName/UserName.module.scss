.userNameContainer {
    display: flex;
    gap: 10px;
    padding: 8px 16px;
    max-width: 300px;
}

.avatarLink {
    width: 32px;
    height: 32px;
}

.avatar {
    width: 32px;
    height: 32px;
    border-radius: 50%;
    border: 1px solid rgb(240, 240, 240);
}

.nicknameWrraper {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    gap: 5px;
    margin: auto 0;
}

.nickname {
    color: rgb(32, 32, 32);
    text-decoration: none;
}

.additionalString {
    color: rgb(142, 142, 142);
    margin: 0;
    font-weight: 200;
    font-size: 12px;
}

.comment {
    display: inline;
    font-weight: 500;
    margin-left: 5px;
    font-size: 14px;
}