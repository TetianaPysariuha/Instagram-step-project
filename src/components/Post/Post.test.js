import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Post from './Post';
import UserName from '../UserName/UserName';
import '@testing-library/jest-dom';

jest.mock('react-redux');
const userName = <UserName />;
const comments = (
  <>
    <p>comment1</p>
    <p>comment2</p>
    <p>comment3</p>
  </>
);

describe('Post snapshot testing', () => {
  test('Should Post to render where isLiked and isFavorite and isMore with value true', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Post
          postId="1"
          userName={userName}
          maneImg="./images/post1.jpeg"
          title="Big Dog and Small Girl"
          isLiked
          handleCklickLike={() => {}}
          handleCklickComments={() => {}}
          isFavorite
          handleCklickFavorite={() => {}}
          comments={comments}
          isMore
          handleCklickShowMore={() => {}}
          handleClickSubmit={() => {}}
        />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Should Post to render where isLiked and isFavorite and isMore with value false', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Post
          postId="1"
          userName={userName}
          maneImg="./images/post1.jpeg"
          title="Big Dog and Small Girl"
          isLiked
          handleCklickLike={() => {}}
          handleCklickComments={() => {}}
          isFavorite
          handleCklickFavorite={() => {}}
          comments={comments}
          isMore
          handleCklickShowMore={() => {}}
          handleClickSubmit={() => {}}
        />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Post buttons work', () => {
  test('Click on button Like', () => {
    const handleCklickLike = jest.fn();
    render(
      <BrowserRouter>
        <Post
          postId="1"
          userName={userName}
          maneImg="./images/post1.jpeg"
          title="Big Dog and Small Girl"
          isLiked
          handleCklickLike={handleCklickLike}
          handleCklickComments={() => {}}
          isFavorite
          handleCklickFavorite={() => {}}
          comments={comments}
          isMore
          handleCklickShowMore={() => {}}
          handleClickSubmit={() => {}}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByTestId('likeBtn'));
    expect(handleCklickLike).toHaveBeenCalledTimes(1);
  });

  test('Click on button comments', () => {
    const handleCklickComments = jest.fn();
    render(
      <BrowserRouter>
        <Post
          postId="1"
          userName={userName}
          maneImg="./images/post1.jpeg"
          title="Big Dog and Small Girl"
          isLiked
          handleCklickLike={() => {}}
          handleCklickComments={handleCklickComments}
          isFavorite
          handleCklickFavorite={() => {}}
          comments={comments}
          isMore
          handleCklickShowMore={() => {}}
          handleClickSubmit={() => {}}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByTestId('commentBtn'));
    expect(handleCklickComments).toHaveBeenCalledTimes(1);
  });

  test('Click on button favorite', () => {
    const handleCklickFavorite = jest.fn();
    render(
      <BrowserRouter>
        <Post
          postId="1"
          userName={userName}
          maneImg="./images/post1.jpeg"
          title="Big Dog and Small Girl"
          isLiked
          handleCklickLike={() => {}}
          handleCklickComments={() => {}}
          isFavorite
          handleCklickFavorite={handleCklickFavorite}
          comments={comments}
          isMore
          handleCklickShowMore={() => {}}
          handleClickSubmit={() => {}}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByTestId('favoriteBtn'));
    expect(handleCklickFavorite).toHaveBeenCalledTimes(1);
  });

  test('Click on button ShowMore', () => {
    const handleCklickShowMore = jest.fn();
    render(
      <BrowserRouter>
        <Post
          postId="1"
          userName={userName}
          maneImg="./images/post1.jpeg"
          title="Big Dog and Small Girl"
          isLiked
          handleCklickLike={() => {}}
          handleCklickComments={() => {}}
          isFavorite
          handleCklickFavorite={() => {}}
          comments={comments}
          isMore
          handleCklickShowMore={handleCklickShowMore}
          handleClickSubmit={() => {}}
        />
      </BrowserRouter>,
    );
    fireEvent.click(screen.getByText('Show more comments', { exact: false }));
    expect(handleCklickShowMore).toHaveBeenCalledTimes(1);
  });
});
