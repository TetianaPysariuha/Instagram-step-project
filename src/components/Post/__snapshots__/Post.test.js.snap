// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post snapshot testing Should Post to render where isLiked and isFavorite and isMore with value false 1`] = `
<DocumentFragment>
  <div
    class="postContainier"
  >
    <div
      class="postWrapper"
    >
      <div
        class="userNameContainer"
      >
        <a
          class="avatarLink"
          href="/undefined/"
        >
          <img
            alt="avatar"
            class="avatar"
            src="default-avatar.jpg"
          />
        </a>
        <div
          class="nicknameWrraper"
        >
          <div>
            <a
              class="nickname"
              href="/undefined/"
            >
              <span />
            </a>
            <p
              class="comment"
            />
          </div>
        </div>
      </div>
      <div
        class="maneFoto"
      >
        <img
          alt="Big Dog and Small Girl"
        />
      </div>
      <div
        class="buttonBox"
      >
        <div
          class="groupBtn"
        >
          <img
            alt="liked"
            data-testid="likeBtn"
            src="./images/heart-red.svg"
          />
          <img
            alt="see all comments"
            data-testid="commentBtn"
            src="./images/comment.svg"
          />
        </div>
        <img
          alt="favorite"
          data-testid="favoriteBtn"
          src="./images/favorite-marked.svg"
        />
      </div>
      <p
        class="description"
      />
      <div
        class="comments"
      >
        <p>
          comment1
        </p>
        <p>
          comment2
        </p>
        <p>
          comment3
        </p>
        <button
          class="showMoreBtn"
          type="button"
        >
          -----Show more comments-----
        </button>
      </div>
      <div
        class="addComment"
      >
        <form
          action="#"
          class="form"
        >
          <input
            class="formField"
            name="comment"
            placeholder="write a comment"
            type="text"
            value=""
          />
          <button
            class="btnSubmit"
            disabled=""
            type="submit"
          >
            Add comment
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Post snapshot testing Should Post to render where isLiked and isFavorite and isMore with value true 1`] = `
<DocumentFragment>
  <div
    class="postContainier"
  >
    <div
      class="postWrapper"
    >
      <div
        class="userNameContainer"
      >
        <a
          class="avatarLink"
          href="/undefined/"
        >
          <img
            alt="avatar"
            class="avatar"
            src="default-avatar.jpg"
          />
        </a>
        <div
          class="nicknameWrraper"
        >
          <div>
            <a
              class="nickname"
              href="/undefined/"
            >
              <span />
            </a>
            <p
              class="comment"
            />
          </div>
        </div>
      </div>
      <div
        class="maneFoto"
      >
        <img
          alt="Big Dog and Small Girl"
        />
      </div>
      <div
        class="buttonBox"
      >
        <div
          class="groupBtn"
        >
          <img
            alt="liked"
            data-testid="likeBtn"
            src="./images/heart-red.svg"
          />
          <img
            alt="see all comments"
            data-testid="commentBtn"
            src="./images/comment.svg"
          />
        </div>
        <img
          alt="favorite"
          data-testid="favoriteBtn"
          src="./images/favorite-marked.svg"
        />
      </div>
      <p
        class="description"
      />
      <div
        class="comments"
      >
        <p>
          comment1
        </p>
        <p>
          comment2
        </p>
        <p>
          comment3
        </p>
        <button
          class="showMoreBtn"
          type="button"
        >
          -----Show more comments-----
        </button>
      </div>
      <div
        class="addComment"
      >
        <form
          action="#"
          class="form"
        >
          <input
            class="formField"
            name="comment"
            placeholder="write a comment"
            type="text"
            value=""
          />
          <button
            class="btnSubmit"
            disabled=""
            type="submit"
          >
            Add comment
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
